libflux/src/semantic/nodes.rs

```diff
@@ -0,0 +1,567 @@
+}
+
+// FunctionExpr represents the definition of a function
+#[derive(Debug, PartialEq, Clone)]
+pub struct FunctionExpr<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub params: Vec<FunctionParameter<'a>>,
+    pub body: FunctionBody<'a>,
+}
+
+impl<'a> FunctionExpr<'a> {
+    pub fn pipe(&self) -> Option<&FunctionParameter<'a>> {
+        for p in &self.params {
+            if p.is_pipe {
+                return Some(p);
+            }
+        }
+        None
+    }
+
+    pub fn defaults(&self) -> Vec<&FunctionParameter<'a>> {
+        let mut ds = Vec::new();
+        for p in &self.params {
+            match p.default {
+                Some(_) => ds.push(p),
+                None => (),
+            }
+        }
+        ds
+    }
+}
+
+// FunctionParameter represents a function parameter.
+#[derive(Debug, PartialEq, Clone)]
+pub struct FunctionParameter<'a> {
+    pub loc: Location<'a>,
+
+    pub is_pipe: bool,
+    pub key: Identifier<'a>,
+    pub default: Option<Expression<'a>>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct BinaryExpr<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub operator: ast::Operator,
+    pub left: Expression<'a>,
+    pub right: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct CallExpr<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub callee: Expression<'a>,
+    pub arguments: Vec<Property<'a>>,
+    pub pipe: Option<Expression<'a>>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ConditionalExpr<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub test: Expression<'a>,
+    pub consequent: Expression<'a>,
+    pub alternate: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct LogicalExpr<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub operator: ast::LogicalOperator,
+    pub left: Expression<'a>,
+    pub right: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct MemberExpr<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub object: Expression<'a>,
+    pub property: String,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct IndexExpr<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub array: Expression<'a>,
+    pub index: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ObjectExpr<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub with: Option<IdentifierExpr<'a>>,
+    pub properties: Vec<Property<'a>>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct UnaryExpr<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub operator: ast::Operator,
+    pub argument: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Property<'a> {
+    pub loc: Location<'a>,
+
+    pub key: PropertyKey<'a>,
+    pub value: Expression<'a>,
```

Comment:
I have updated this not to be optional (I realized that that was useless as you noted).